testament/testament.nim

```diff
@@ -2017,6 +2080,17 @@ proc runTests(execState: var Execution) =
   startTimes = newSeq[float](batchSize)
   endTimes = newSeq[float](batchSize)
 
+  # REFACTOR: this function requires additional context to be available,
+  # making it harder to move out of body. It mutates some of the state as
+  # well, so one possible solution would be to make the surrounding state
+  # of the `runTest` function *explicit* -- create a private
+  # `RunTestContext` ref object and then create a `initOnTestRunStartCb`
+  # which takes a context and returns a closure.
+  #
+  # NOTE: This approach might also address the [[mutation]] question above,
```

Comment:
Hmm, I think capturing the test result data (things that the callbacks mutate) in a temporary spot, like a `TestBatch` or whatever, and then integrating it back into the main execution state as a separate step might be ok. Though, I worry it might simply be more code where things can go wrong and not really buying us anything.

Reply:
I don't think there is an explanation in terms of some higher-level abstractions, but when I see the 15-argument with half arguments mutable it feels like a code smell